proxy/src/conditional.rs

```diff
@@ -0,0 +1,56 @@
+use std;
+
+/// Like `std::option::Option<C>` but `None` carries a reason why the value
+/// isn't available.
```

Comment:
I also don't like the introduction of `Conditional` much but this shouldn't be `Result` since `Result` should be used strictly for errors. `Either` makes more sense but I don't want to import another library. Note that this is basically your `Either` type that you implemented `Stream` on. We could move `Either` to a more generic place and then use that instead?

Reply:
> this shouldn't be `Result` since `Result` should be used strictly for errors.

I think "a reason we were not able to do TLS" falls into a (sufficiently broad) definition of "error"; `Err` is used elsewhere for signalling that something is not available. However, I'm not going to press that because I think the discussion is just semantics at this point. :)

> We could move Either to a more generic place and then use that instead?

I would be fine with that as well, but (coming at it from another direction) I think `Conditional` does better encode the meaning of each branch, so, having thought about it a bit, I'm ready to put a green check on this as it is.